### PR TITLE
Separate move path tracking between borrowck and drop elaboration.

### PR DESCRIPTION
**Pull request by cjgillot**
The primary goal of this PR is to skip creating a `MovePathIndex` for path that do not need dropping in drop elaboration.

The 2 first commits are cleanups.

The next 2 commits displace `move` errors from move-path builder to borrowck. Move-path builder keeps the same logic, but does not carry error information any more.

The remaining commits allow to filter `MovePathIndex` creation according to types. This is used in drop elaboration, to avoid computing dataflow for paths that do not need dropping.

### CODE REVIEW

**Comment by cjgillot** [comment]
r? compiler

**Comment by petrochenkov** [comment]
I'm not very familiar with this code, but I worked with move paths several years ago when implementing unions, and I need to get familiar with it again for my work, so I'll review anyway.

File: compiler/rustc_borrowck/src/diagnostics/move_errors.rs [hunk 1]
```diff
@@ -1,16 +1,50 @@
...
+use rustc_mir_dataflow::move_paths::{LookupResult, MovePathIndex};
 use rustc_span::{BytePos, Span};
 
 use crate::diagnostics::CapturedMessageOpt;
 use crate::diagnostics::{DescribePlaceOpt, UseSpans};
 use crate::prefixes::PrefixSet;
 use crate::MirBorrowckCtxt;
 
+#[derive(Debug)]
+pub enum IllegalMoveOriginKind<'tcx> {
```
Comment (petrochenkov) [inline compiler/rustc_borrowck/src/diagnostics/move_errors.rs:13]: In this commit it's hard to see that `IllegalMoveOriginKind` is not a new structure but is actually copypasted from elsewhere, and that the copy is identical.
Maybe it should initially be imported instead.

File: compiler/rustc_borrowck/src/lib.rs [hunk 2]
```diff
@@ -1465,6 +1470,9 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                 );
             }
             Operand::Move(place) => {
+                // Check if moving from this place makes sense.
+                self.check_movable_place(location, place);
```
Comment (petrochenkov) [inline compiler/rustc_borrowck/src/lib.rs:1476]: Why is `check_movable_place` called from two places?
Can it be centralized to one place?

File: compiler/rustc_borrowck/src/lib.rs [hunk 3]
```diff
@@ -1586,6 +1594,142 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+                            },
+                        ));
+                        return;
+                    }
+                    _ => bug!("Unexpected type {place_ty:#?}"),
+                },
+                // `OpaqueCast` only transmutes the type, so no moves there and
+                // `Downcast` only changes information about a `Place` without moving
+                // So it's safe to skip these.
+                ProjectionElem::OpaqueCast(_) | ProjectionElem::Downcast(_, _) => (),
```
Comment (petrochenkov) [inline compiler/rustc_borrowck/src/lib.rs:1728]: Is it possible to share code between this function and `fn move_path_for` somehow?

In this commit it's hard to see that this logic is not new but is actually copied from elsewhere, and that the logic is identical in those two places.

File: compiler/rustc_borrowck/src/lib.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_borrowck/src/lib.rs:1476]: We want to report move errors from promoted, but the different borrowck analyses are still run manually for them. This is the least-invasive way to keep the behaviour.
The "proper" way to unify both cases would be https://github.com/rust-lang/rust/pull/111693 or something like it.

File: compiler/rustc_borrowck/src/lib.rs (hunk 3 above)
Comment (petrochenkov) [reply compiler/rustc_borrowck/src/lib.rs]: Generic version, if that's not too ugly.

File: compiler/rustc_mir_transform/src/elaborate_drops.rs [hunk 4]
```diff
@@ -324,80 +334,74 @@ impl<'b, 'tcx> ElaborateDropsCtxt<'b, 'tcx> {
...
-                            elaborate_drop(
-                                &mut Elaborator { ctxt: self },
-                                terminator.source_info,
-                                place,
-                                path,
-                                target,
-                                unwind,
-                                bb,
-                            )
+            // This place does not need dropping. It is not have an associated move-path, so the
```
Comment (petrochenkov) [inline compiler/rustc_mir_transform/src/elaborate_drops.rs:350]:
```suggestion
            // This place does not need dropping. It does not have an associated move-path, so the
```

**Comment by petrochenkov** [comment]
r=me after addressing the remaining comments (https://github.com/rust-lang/rust/pull/116300#discussion_r1355119446, https://github.com/rust-lang/rust/pull/116300#discussion_r1357296014) or confirming that they are not worth addressing.
@rustbot author

File: compiler/rustc_borrowck/src/lib.rs (hunk 3 above)
Comment (cjgillot) [reply compiler/rustc_borrowck/src/lib.rs]: I don't think keeping a common version is an interesting path forward. The borrowck version is tailored to the language behaviour in borrowck, ie user-facing. The mir_dataflow version could evolve into being more syntactic, and check less types.

**Pull request merged**